docs/architecture/adr-008-app-caller-cbs.md

```diff
@@ -0,0 +1,526 @@
+		ctx, packet.GetSourcePort(), packet.GetSourceChannel(), packet.GetSequence(),
+		types.CallbackTypeTimeoutPacket, callbackData, err,
+	)
+
+	return nil
+}
+
+// processCallback executes the callbackExecutor and reverts contract changes if the callbackExecutor fails.
+//
+// Error Precedence and Returns:
+//   - oogErr: Takes the highest precedence. If the callback runs out of gas, an error wrapped with types.ErrCallbackOutOfGas is returned.
+//   - panicErr: Takes the second-highest precedence. If a panic occurs and it is not propagated, an error wrapped with types.ErrCallbackPanic is returned.
+//   - callbackErr: If the callbackExecutor returns an error, it is returned as-is.
+//
+// panics if
+//   - the contractExecutor panics for any reason, and the callbackType is SendPacket, or
+//   - the contractExecutor runs out of gas and the relayer has not reserved gas grater than or equal to
+//     CommitGasLimit.
+func (IBCMiddleware) processCallback(
+	ctx sdk.Context, callbackType types.CallbackType,
+	callbackData types.CallbackData, callbackExecutor func(sdk.Context) error,
+) (err error) {
+	cachedCtx, writeFn := ctx.CacheContext()
+	cachedCtx = cachedCtx.WithGasMeter(storetypes.NewGasMeter(callbackData.ExecutionGasLimit))
+
+	defer func() {
+		// consume the minimum of g.consumed and g.limit
+		ctx.GasMeter().ConsumeGas(cachedCtx.GasMeter().GasConsumedToLimit(), fmt.Sprintf("ibc %s callback", callbackType))
+
+		// recover from all panics except during SendPacket callbacks
+		if r := recover(); r != nil {
+			if callbackType == types.CallbackTypeSendPacket {
+				panic(r)
+			}
+			err = errorsmod.Wrapf(types.ErrCallbackPanic, "ibc %s callback panicked with: %v", callbackType, r)
+		}
+
+		// if the callback ran out of gas and the relayer has not reserved enough gas, then revert the state
+		if cachedCtx.GasMeter().IsPastLimit() {
+			if callbackData.AllowRetry() {
+				panic(storetypes.ErrorOutOfGas{Descriptor: fmt.Sprintf("ibc %s callback out of gas; commitGasLimit: %d", callbackType, callbackData.CommitGasLimit)})
+			}
+			err = errorsmod.Wrapf(types.ErrCallbackOutOfGas, "ibc %s callback out of gas", callbackType)
+		}
+
+		// allow the transaction to be committed, continuing the packet lifecycle
+	}()
+
+	err = callbackExecutor(cachedCtx)
+	if err == nil {
+		writeFn()
+	}
+
+	return err
+}
+```
+
+Chains are expected to specify a `chainDefinedActorCallbackLimit` to ensure that callbacks do not consume an arbitrary amount of gas. Thus, it should always be possible for a relayer to complete the packet lifecycle even if the actor callbacks cannot run successfully.
+
+## Consequences
+
+### Positive
+
+- IBC Actors can now programatically execute logic that involves sending a packet and then performing some additional logic once the packet lifecycle is complete
+- Middleware implementing ADR-8 can be generally used for any application
+- Leverages the same callback architecture used between core IBC and IBC applications
```

Comment:
```suggestion
- Leverages a similar callback architecture to the one used between core IBC and IBC applications
```

Reply:
why similar and not same?

Reply:
Because they don't use the same interfaces to make the callback (which would be `IBCModule`). Instead, we use `ContractKeeper` which is missing some channel lifetime event callbacks and some parameters. 

Also, the IBC application callback erroring out means that the packet lifecycle is blocked. This is not the case for adr8